lib/maths/CCalendarCyclicTest.cc

```diff
@@ -0,0 +1,328 @@
+        double r{static_cast<double>(stat.second.s_Repeats)};
+        double x{stat.second.s_Count};
+        double e{stat.second.s_Sum};
+        double s{stat.second.s_Significance};
+        if (stat.second.s_Repeats >= MINIMUM_REPEATS && e > errorThreshold * x &&
+            std::pow(s, r) < MAXIMUM_SIGNIFICANCE) {
+            result.add({e, stat.second.s_Offset, feature});
+        }
+    }
+
+    return result.count() > 0 ? result[0].third : TOptionalFeature();
+}
+
+std::uint64_t CCalendarCyclicTest::checksum(std::uint64_t seed) const {
+    seed = CChecksum::calculate(seed, m_DecayRate);
+    seed = CChecksum::calculate(seed, m_ErrorQuantiles);
+    seed = CChecksum::calculate(seed, m_CurrentBucketTime);
+    seed = CChecksum::calculate(seed, m_CurrentBucketIndex);
+    seed = CChecksum::calculate(seed, m_CurrentBucketErrorStats);
+    TErrorStatsVec errors{this->inflate()};
+    return CChecksum::calculate(seed, errors);
+}
+
+void CCalendarCyclicTest::debugMemoryUsage(core::CMemoryUsage::TMemoryUsagePtr mem) const {
+    mem->setName("CCalendarCyclicTest");
+    core::CMemoryDebug::dynamicSize("m_ErrorQuantiles", m_ErrorQuantiles, mem);
+    core::CMemoryDebug::dynamicSize("m_CompressedBucketErrorStats",
+                                    m_CompressedBucketErrorStats, mem);
+}
+
+std::size_t CCalendarCyclicTest::memoryUsage() const {
+    std::size_t mem{core::CMemory::dynamicSize(m_ErrorQuantiles)};
+    mem += core::CMemory::dynamicSize(m_CompressedBucketErrorStats);
+    return mem;
+}
+
+double CCalendarCyclicTest::winsorise(double error) const {
+    double high;
+    m_ErrorQuantiles.quantile(99.5, high);
+    return std::min(error, high);
+}
+
+double CCalendarCyclicTest::significance(double n, double x) const {
+    if (n > 0.0) {
+        try {
+            // We have roughly 31 independent error samples, one for each
+            // day of the month, so the chance of seeing as extreme an event
+            // among all of them is:
+            //   1 - P("don't see as extreme event") = 1 - (1 - P("event"))^31
+            boost::math::binomial binom{n, 1.0 - LARGE_ERROR_PERCENTILE / 100.0};
+            double p{std::min(2.0 * CTools::safeCdfComplement(binom, x - 1.0), 1.0)};
+            return CTools::oneMinusPowOneMinusX(p, 31.0);
+        } catch (const std::exception& e) {
+            LOG_ERROR(<< "Failed to calculate significance: " << e.what()
+                      << " n = " << n << " x = " << x);
+        }
+    }
+    return 1.0;
+}
+
+void CCalendarCyclicTest::deflate(const TErrorStatsVec& stats) {
+    bool lengthOnly{false};
+    core::CDeflator deflator{lengthOnly};
+    deflator.addVector(stats);
+    deflator.finishAndTakeData(m_CompressedBucketErrorStats);
+    m_CompressedBucketErrorStats.shrink_to_fit();
+}
+
+CCalendarCyclicTest::TErrorStatsVec CCalendarCyclicTest::inflate() const {
+    bool lengthOnly{false};
+    core::CInflator inflator{lengthOnly};
```

Comment:
I wonder if there is room for improvement here, instantiating Inflator/Deflator is quite expensive and adds to heap fragmentation (would be good to do a long running test to be sure this isn't a problem). We could have some compression/decompression service which internally reuses buffers. 

Nit?: In terms of memory usage: It sounds a bit like cheating to me, we are not accounting for the memory that is temporary used. I understand that this isn't a big deal if you have a lot of partitions for small jobs it adds 4K++ which are not accounted anymore.

Reply:
> happen approximately once per day

Ok, in this case the comment above is probably solved, I did not see this running only once per day.

Reply:
Am I correct that it's once per day _per time series_?

It also happens in persist/restore, which is an extra 6-8 times per day per time series at the default background persist frequency.

But it depends on whether creating and destroying the zlib data structures is really significant compared to all the other processing we're doing.  A shared service would need to consider thread safety, maybe by caching an inflator and deflator per thread, so if the overhead isn't measurable compared to the other CPU-intensive processing we're doing then it may not be worth the added complexity.

Reply:
agreed, I think it only makes sense to think about a service if it's used by more than this, which is the case: "another part of the model state which benefits from storing some of its state in a compressed format."

Having that said, before looking into that, I would 1st check zlib alternatives, e.g. snappy: It's faster, it does not require initialization/deinitialization (-> therefore a service is superfluous). But it might be a little bit weaker in compression (needs measurement!).

^ But **not** important, just an idea for a potential future improvement which might be of interest if we use compression in more places.

Reply:
Regarding runtime, you're right strictly it's `1 + (persists + restores) x "time series"` per day.

I profiled running the executable standalone for the "worst case scenario": polled, at the bucket length (= 5m), normally distributed data with many series. This gives us about the simplest possible modelling and the minimum possible overhead for processing the raw data to extract bucket statistics. I'd say this is pathological, i.e. I don't think it is realistic, but even so the cost still only comes in at about 4% increase in run time due to the compress/decompress. That also excludes any time spent in Java. So I think this will not introduce a new CPU bottleneck.

Since this solution is performant enough I think I'd prefer to just stick with zlib and temporaries. (I agree it would be nice to avoid memory churn, but I think it make sense to wait for the C++17 standard library enhancements, when we could revisit allocation strategies in general.)

Regarding memory usage, there is as much (and often more) memory not accounted for elsewhere, for example in `api::CAnomalyJob` and in temporaries we create, principally the results object. In terms of job allocation we have a pad for a single job which I think more than covers this. Since it is fixed,  independent of the number of time series modelled, I think the small additional temporary memory this PR introduces is ok.

I wouldn't be against trying to get any memory accounting we do better and in particular to estimate peak usage. I think my preferred solution would be an empirically determine a margin, but I'd prefer to keep that as a separate change.